go/pkg/cosmos/gas.go

```diff
@@ -0,0 +1,47 @@
+package cosmos
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"strconv"
+
+	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/pkg/errors"
+)
+
+func (c *HTTPClient) GetEstimateGas(txBytes []byte) (string, error) {
+	var res txtypes.SimulateResponse
+	type retStatus struct {
+		Code   uint32        `json:"code"`
+		Msg    string        `json:"message"`
+		Detail []interface{} `json:"detail"`
+	}
+	reqData, err := base64.StdEncoding.DecodeString(string(txBytes))
+	if err != nil {
+		return "", errors.Wrapf(err, "error decoding transaction from base64")
+	}
+	reqRawBody := txtypes.SimulateRequest{TxBytes: reqData}
+	jsonBody, err := c.encoding.Marshaler.MarshalJSON(&reqRawBody)
+	if err != nil {
+		return "", err
+	}
+
+	ret, err := c.cosmos.R().SetBody(jsonBody).SetResult(&res).Post("/cosmos/tx/v1beta1/simulate")
+	if err != nil || ret.StatusCode() != 200 {
+		var errorMsg = retStatus{}
+		json.Unmarshal(ret.Body(), &errorMsg)
+		return errorMsg.Msg, errors.Wrapf(err, errorMsg.Msg)
+	}
+
+	return strconv.FormatUint(res.GasInfo.GasUsed, 10), nil
+}
```

Comment:
```suggestion
func (c *HTTPClient) GetEstimateGas(rawTx string) (string, error) {
	txBytes, err := base64.StdEncoding.DecodeString(rawTx)
	if err != nil {
		return "", errors.Wrapf(err, "failed to decode rawTx: %s", rawTx)
	}

	e := &ErrorResponse{}
	res := &txtypes.SimulateResponse{}

	_, err = c.cosmos.R().SetBody(txtypes.SimulateRequest{TxBytes: txBytes}).SetResult(res).SetError(e).Post("/cosmos/tx/v1beta1/simulate")
	if err != nil {
		return "", errors.Wrap(err, "failed to estimate gas")
	}

	if e != nil {
		return "", errors.Errorf("failed to estimate gas: %s", e.Msg)
	}

	return strconv.FormatUint(res.GasInfo.GasUsed, 10), nil
}
```

Reply:
Can I ask where is the `&ErrorResponse{}`??

Reply:
```
diff --git a/go/pkg/cosmos/types.go b/go/pkg/cosmos/types.go
index 21c854d..d2fdbfc 100644
--- a/go/pkg/cosmos/types.go
+++ b/go/pkg/cosmos/types.go
@@ -43,6 +43,13 @@ type Delegation struct {
 	Balance Value `json:"balance"`
 }
 
+// ErrorResponse payload for an api request
+type ErrorResponse struct {
+	Code   int           `json:"code"`
+	Msg    string        `json:"message"`
+	Detail []interface{} `json:"detail"`
+}
+
 // Contains info about a transaction log event
 // swagger:model Event
 type Event struct {
```